src/utils/keyAndAdresses.ts

```diff
@@ -0,0 +1,71 @@
+import {
+  base58Decode,
+  base58Encode,
+  varintEncode,
+  varintDecode,
+  hashBlake3,
+} from './Xbqcrypto';
+
+import { getBytesSecretKeyVersioned } from './bytes';
+
+const PUBLIC_KEY_PREFIX = 'P';
+const ADDRESS_PREFIX = 'AU';
+const SECRET_KEY_PREFIX = 'S';
+
+export class SecretKey {
+  base58Encoded: string;
+  base58Decoded: Uint8Array;
+  prefix: string;
+  version: number;
+
+  constructor(secretKeyBase58Encoded: string) {
+    this.prefix = SECRET_KEY_PREFIX;
+    this.base58Encoded = secretKeyBase58Encoded;
+    this.base58Decoded = getBytesSecretKeyVersioned(secretKeyBase58Encoded);
+
+    const secretKeyUnprefixed = secretKeyBase58Encoded.slice(
+      SECRET_KEY_PREFIX.length,
+    );
+    const decoded = base58Decode(secretKeyUnprefixed);
+    const decodedVersion = varintDecode(decoded);
+    this.version = decodedVersion.value;
+  }
+}
+
+export class PublicKey {
+  version: number;
+  publicKey: Uint8Array;
+  base58Encoded: string;
+
+  constructor(publicKeyArray: Uint8Array, secretKey: SecretKey) {
+    this.version = secretKey.version;
```

Comment:
By creating the public key from the secret key object, it requires to have the same version for the public key and secret key. Thus, we can't have different versions.
We could do a different constructor with in arg the publicKeyBase58encoded, and so we could decode the version, but sometimes we don't have the publicKeyBase58encoded yet